### PR TITLE
refactor: Refactor DateTypes to make it more easy to use

### PR DESCRIPTION
**Pull request by Xuanwo**
This PR refactor data types to make it more easy to use in rust.

Some `Display` doesn't implemented correctly, I removed them in this PR. We can add them back in following PRs.

### CODE REVIEW

**Review by XuQianJin-Stars — APPROVED** [review]
LGTM

**Review by devillove084 — APPROVED** [review]

File: crates/paimon/src/spec/types.rs [hunk 1]
```diff
@@ -329,44 +225,40 @@ impl BinaryType {
 
     pub const DEFAULT_LENGTH: usize = 1;
 
-    pub fn new(is_nullable: bool, length: usize) -> Self {
-        Self::try_new(is_nullable, length).unwrap()
+    pub fn new(length: usize) -> Result<Self, &'static str> {
```
[the remark this answers is not in the document]
Comment (Xuanwo) [reply crates/paimon/src/spec/types.rs:228]: > You might consider unifying the error types, such as using InvalidXXXType, and applying the same approach in other areas as well.

Yes, we've directly migrated the code here without refactoring errors to clarify the scope of this PR. We can finish this part later.

[the remark this answers is not in the document]
Comment (Xuanwo) [reply crates/paimon/src/spec/types.rs:228]: Would you like to perform this refactor? The error type has been defined at https://github.com/apache/paimon-rust/blob/a5cfac79ae4e9eaf87b752b3720fb0575cef3cbb/crates/paimon/src/error.rs#L25-L40

File: crates/paimon/src/spec/types.rs [hunk 2]
```diff
@@ -164,118 +112,55 @@ impl FromStr for DataType {
...
-        Self {
-            is_nullable,
-            type_root: self.type_root,
-        }
-    }
-
-    /// Returns true if the data type is nullable.
-    ///
-    /// Impl Reference: <https://github.com/apache/paimon/blob/release-0.8.2/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L59>
-    fn is_nullable(&self) -> bool {
```
Comment (Aitozi) [inline crates/paimon/src/spec/types.rs:188]: Is this method should be kept for the `DataType` ?

Comment (Xuanwo) [reply crates/paimon/src/spec/types.rs:188]: Yes, we can implement in following PRs.

**Review by Aitozi — APPROVED** [review]
+1

File: crates/paimon/src/spec/types.rs (hunk 2 above)
Comment (Aitozi) [reply crates/paimon/src/spec/types.rs:188]: OK

Comment (Xuanwo) [reply crates/paimon/src/spec/types.rs:188]: Tracked at https://github.com/apache/paimon-rust/issues/31

File: crates/paimon/src/spec/types.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xuanwo) [reply crates/paimon/src/spec/types.rs:228]: Tracked at https://github.com/apache/paimon-rust/issues/32

**Pull request merged**